jax/_src/scipy/signal.py

```diff
@@ -146,3 +148,351 @@ def detrend(data, axis=-1, type='linear', bp=0, overwrite_data=None):
+    func = jax.numpy.fft.fft
+  else:
+    result = result.real
+    func = jax.numpy.fft.rfft
+  result = func(result, n=nfft)
+
+  return result
+
+@_wraps(osp_signal.signaltools.odd_ext)
+def odd_ext(x, n, axis=-1):
+  """Redefined with jax.numpy for supporting auto-diff."""
+  def axis_slice(a, start=None, stop=None, step=None, axis=-1):
+    a_slice = [slice(None)] * a.ndim
+    a_slice[axis] = slice(start, stop, step)
+    b = a[tuple(a_slice)]
+    return b
+  if n < 1:
+    return x
+  if n > x.shape[axis] - 1:
+    raise ValueError(("The extension length n (%d) is too big. " +
+                      "It must not exceed x.shape[axis]-1, which is %d.")
+                      % (n, x.shape[axis] - 1))
+  left_end = axis_slice(x, start=0, stop=1, axis=axis)
+  left_ext = axis_slice(x, start=n, stop=0, step=-1, axis=axis)
+  right_end = axis_slice(x, start=-1, axis=axis)
+  right_ext = axis_slice(x, start=-2, stop=-(n + 2), step=-1, axis=axis)
+  ext = jnp.concatenate((2 * left_end - left_ext,
+                         x,
+                         2 * right_end - right_ext),
+                         axis=axis)
+  return ext
+
+
+_spectral_helper_mode = ('psd', 'stft')
+
+def _spectral_helper(x, y, fs=1.0, window='hann', nperseg=None, noverlap=None,
+                     nfft=None, detrend_type='constant', return_onesided=True,
+                     scaling='density', axis=-1, mode='psd', boundary=None,
+                     padded=False):
+  if mode not in _spectral_helper_mode:
+    raise ValueError(f"Unknown value for mode {mode}, "
+                     f"must be one of: {_spectral_helper_mode}")
+
+  def make_pad(mode, **kwargs):
+    def pad(x, n, axis=-1):
+      pad_width = [(0, 0) for unused_n in range(x.ndim)]
+      pad_width[axis] = (n, n)
+      return jnp.pad(x, pad_width, mode, **kwargs)
+    return pad
+
+  boundary_funcs = {
+      'even': make_pad('reflect'),
+      'odd': odd_ext,
+      'constant': make_pad('edge'),
+      'zeros': make_pad('constant', constant_values=0.0),
+      None: lambda x, *args, **kwargs: x
+  }
+
+  if boundary not in boundary_funcs:
+    raise ValueError(
+        f"Unknown boundary option '{boundary}', "
+        f"must be one of: {list(boundary_funcs.keys())}")
+
+  # If x and y are the same object we can save ourselves some computation.
+  same_data = y is x
+
+  if not same_data and mode != 'psd':
+    raise ValueError("x and y must be equal if mode is not 'psd'")
+
+  axis = int(axis)
+
+  # Ensure we have np.arrays, get outdtype
+  x = jnp.asarray(x)
+  if not same_data:
+    y = jnp.asarray(y)
+    outdtype = jnp.result_type(x, y, np.complex64)
+  else:
+    outdtype = jnp.result_type(x, np.complex64)
+
+  if not same_data:
+    # Check if we can broadcast the outer axes together
+    xouter = list(x.shape)
+    youter = list(y.shape)
+    xouter.pop(axis)
+    youter.pop(axis)
+    try:
+      outershape = jnp.broadcast_arrays(
+          np.empty(xouter), jnp.empty(youter))[0].shape
```

Comment:
We can use `lax.broadcast_shapes` for this, which will avoid creating the broadcasted versions of the arrays.

Reply:
changed to use `jnp.broadcast_shapes`.